packages/remix-dev/config/flat-routes.ts

```diff
@@ -0,0 +1,269 @@
+
+import type { ConfigRoute, RouteManifest } from "./routes";
+import {
+  escapeEnd,
+  escapeStart,
+  isSegmentSeparator,
+  optionalEnd,
+  optionalStart,
+  paramPrefixChar,
+  routeModuleExts,
+} from "./routesConvention";
+
+export function flatRoutes(
+  appDirectory: string,
+  ignoredFilePatterns?: string[]
+): RouteManifest {
+  let extensions = routeModuleExts.join(",");
+
+  let routePaths = fg.sync(`*{${extensions}}`, {
+    absolute: true,
+    cwd: path.join(appDirectory, "routes"),
+    ignore: ignoredFilePatterns,
+  });
+
+  return flatRoutesUniversal(appDirectory, routePaths);
+}
+
+/**
+ * Create route configs from a list of routes using the flat routes conventions.
+ * @param {string} appDirectory - The absolute root directory the routes were looked up from.
+ * @param {string[]} routePaths - The absolute route paths.
+ * @param {string} [prefix=routes] - The prefix to strip off of the routes.
+ */
+export function flatRoutesUniversal(
+  appDirectory: string,
+  routePaths: string[],
+  prefix: string = "routes"
+): RouteManifest {
+  let routes: Record<string, ConfigRoute> = {};
+  let sortedRoutes = routePaths
+    .sort((a, b) => (a.length - b.length > 0 ? 1 : -1))
+    .map((routePath) =>
+      routePath.slice(appDirectory.length + 1).replace(/\\/g, "/")
+    );
+  let processedIds: string[] = [];
+  for (let routePath of sortedRoutes) {
+    let routeId = routeIdFromPath(routePath);
+    if (routes[routeId]) {
+      throw new Error(`Duplicate route: ${routeId}`);
+    }
+
+    let tmpId = "";
+    let parentId = "";
+    for (let processedId of processedIds) {
+      if (routeId.startsWith(processedId.replace(/\$$/, "*"))) {
+        let [segments, delimiters] = splitByDelimiter(routeId);
+        if (segments.at(1)?.endsWith("_")) {
+          segments[1] = segments[1].slice(0, -1);
+          tmpId = joinByDelimiter(segments, delimiters);
+          break;
+        }
+
+        parentId = processedId;
+        break;
+      }
+    }
+
+    routes[routeId] = {
+      id: routeId,
+      path: pathFromRouteId(tmpId || routeId, parentId || prefix),
+      parentId: parentId || "root",
+      file: routePath,
+    };
+    if (isIndexRoute(routeId)) {
+      routes[routeId].index = true;
+    }
+    processedIds.unshift(routeId);
+  }
+
+  return routes;
+}
+
+function routeIdFromPath(relativePath: string) {
+  return (
+    relativePath
+      .split(".")
+      // remove file extension
+      .slice(0, -1)
+      .join(".")
+  );
+}
+
+function pathFromRouteId(routeId: string, parentId: string) {
+  let parentPath = "";
+  if (parentId) {
+    parentPath = getRouteSegments(parentId).join("/");
+  }
+  if (parentPath.startsWith("/")) {
+    parentPath = parentPath.substring(1);
+  }
+  let routePath = getRouteSegments(routeId).join("/");
+  if (routePath.startsWith("/")) {
+    routePath = routePath.substring(1);
+  }
+  let pathname = parentPath
+    ? routePath.slice(parentPath.length + 1)
+    : routePath;
```

Comment:
Suggestion: could use [lodash's `escapeRegExp`](https://lodash.com/docs/4.17.15#escapeRegExp) to be more explicit:

```suggestion
  let pathname = routePath.replace(new RegExp("^" + _.escapeRegExp(parentPath)), '')
```

Reply:
code no longer exists in this form